tfhe/src/integer/server_key/comparator.rs

```diff
@@ -906,6 +906,246 @@ impl<'a> Comparator<'a> {
+                rayon::join(
+                    || self.server_key.full_propagate_parallelized(&mut tmp_lhs),
+                    || self.server_key.full_propagate_parallelized(&mut tmp_rhs),
+                );
+                (&tmp_lhs, &tmp_rhs)
+            }
+        };
+
+        self.unchecked_lt_parallelized(lhs, rhs)
+    }
+
+    pub fn le_parallelized<PBSOrder: PBSOrderMarker>(
+        &self,
+        lhs: &RadixCiphertext<PBSOrder>,
+        rhs: &RadixCiphertext<PBSOrder>,
+    ) -> RadixCiphertext<PBSOrder> {
+        let mut tmp_lhs: RadixCiphertext<PBSOrder>;
+        let mut tmp_rhs: RadixCiphertext<PBSOrder>;
+        let (lhs, rhs) = match (lhs.block_carries_are_empty(), rhs.block_carries_are_empty()) {
+            (true, true) => (lhs, rhs),
+            (true, false) => {
+                tmp_rhs = rhs.clone();
+                self.server_key.full_propagate_parallelized(&mut tmp_rhs);
+                (lhs, &tmp_rhs)
+            }
+            (false, true) => {
+                tmp_lhs = lhs.clone();
+                self.server_key.full_propagate_parallelized(&mut tmp_lhs);
+                (&tmp_lhs, rhs)
+            }
+            (false, false) => {
+                tmp_lhs = lhs.clone();
+                tmp_rhs = rhs.clone();
+                rayon::join(
+                    || self.server_key.full_propagate_parallelized(&mut tmp_lhs),
+                    || self.server_key.full_propagate_parallelized(&mut tmp_rhs),
+                );
+                (&tmp_lhs, &tmp_rhs)
+            }
+        };
+
+        self.unchecked_le_parallelized(lhs, rhs)
+    }
+
+    pub fn max_parallelized<PBSOrder: PBSOrderMarker>(
+        &self,
+        lhs: &RadixCiphertext<PBSOrder>,
+        rhs: &RadixCiphertext<PBSOrder>,
+    ) -> RadixCiphertext<PBSOrder> {
+        let mut tmp_lhs: RadixCiphertext<PBSOrder>;
+        let mut tmp_rhs: RadixCiphertext<PBSOrder>;
+
+        let (lhs, rhs) = match (lhs.block_carries_are_empty(), rhs.block_carries_are_empty()) {
+            (true, true) => (lhs, rhs),
+            (true, false) => {
+                tmp_rhs = rhs.clone();
+                self.server_key.full_propagate_parallelized(&mut tmp_rhs);
+                (lhs, &tmp_rhs)
+            }
+            (false, true) => {
+                tmp_lhs = lhs.clone();
+                self.server_key.full_propagate_parallelized(&mut tmp_lhs);
+                (&tmp_lhs, rhs)
+            }
+            (false, false) => {
+                tmp_lhs = lhs.clone();
+                tmp_rhs = rhs.clone();
+                rayon::join(
+                    || self.server_key.full_propagate_parallelized(&mut tmp_lhs),
+                    || self.server_key.full_propagate_parallelized(&mut tmp_rhs),
+                );
+                (&tmp_lhs, &tmp_rhs)
+            }
+        };
+
+        let mut res = self.unchecked_max_parallelized(lhs, rhs);
+        self.server_key.full_propagate_parallelized(&mut res);
```

Comment:
I think we can do better than full_propagate.

The last step of unchecked_max/min is an unchecked_add(x,y)
where 
- if x != 0 => 0 < x < msg_mod and y == 0
- or if y != 0 => x == 0 and  0 < y < msg_mod

so the resulting ciphertext should not have any carries, 
thus the degree is in reality  message_mod - 1,
which means that is we want to reset the degree/noise we whould simply `clean_carry` in parallel

Reply:
My test was failing I believe when I did not do the full propagate for min and max

Reply:
ah right, you may be correct here

Reply:
on my side, without the pull propagate the test passes

but we might do:
`res.blocks.par_iter_mut().for_each(|block| self.server_key.key.message_extract_assign(block));`
to reset the noise

Reply:
I think it failed in the CI here, I relaunched it several times locally and the full propagate was necessary (for the empty carry check I believe), I'll see for the message_extract_assign

Reply:
Since the unchecked_add updates the degree, if you check the degree of the block they will be == (2 * (message_mod - 1)) 
when in reality they should be at message_mod - 1

Reply:
yep, but the noise refresh is probably necessary as degree and noise were managed together, the good news is the refresh can be threaded which was not the case of the propagate

Reply:
doesn't work, will have to full propagate for 3_3 at least

Reply:
so it works for 2_2, 4_4 but not 3_3 ?